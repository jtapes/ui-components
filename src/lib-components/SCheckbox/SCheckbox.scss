@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-checkbox {
  display: inline-flex;
  align-items: center;
  cursor: pointer;

  &__input {
    position: relative;
    display: inline-flex;
    flex-shrink: 0;
    margin-top: $size--4;
    width: $size--16;
    height: $size--16;
  }

  &__inner {
    position: absolute;
    width: 100%;
    height: 100%;
    color: $color--white;
    background-color: $color--primary;
    transform: scaleY(0);
    transition: $animate--time;
  }

  &__original {
    appearance: none;
    border: 1px solid $color--font-muted;
    width: 100%;
    height: 100%;
    transition: $animate--time;
    cursor: pointer;

    &:hover {
      border-color: $color--primary-accent;
    }
  }

  &__label {
    overflow: hidden;
    margin-left: $size--8;
  }

  &--checked & {
    &__label {
      color: $color--primary;
    }

    &__inner {
      transform: scaleY(1);
    }
  }

  &--disabled {
    pointer-events: none;

    .s-checkbox {
      &__label {
        color: $color--font-muted;
      }

      &__original,
      &__inner {
        border-color: $color--font-muted;
        background-color: $color--font-muted;
      }
    }
  }

  &:not(&--disabled):active &__original {
    border-color: transparent;
    box-shadow: 0 0 1px 2px $color--outline;
  }
}
