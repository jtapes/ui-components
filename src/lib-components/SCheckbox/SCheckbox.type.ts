export type SCheckboxValue = object | number | string | boolean
export enum SCheckboxStates {
  DISABLED = 'disabled',
  ERROR = 'error',
  DEFAULT = 'default'
}
