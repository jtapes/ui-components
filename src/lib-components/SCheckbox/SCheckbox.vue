<template>
  <label
    class="s-checkbox text-sm"
    :class="[
      `s-checkbox--${state}`,
      {
        's-checkbox--checked': isChecked,
      },
    ]"
  >
    <div class="s-checkbox__input">
      <div class="s-checkbox__inner">
        <svg-icon name="checked-16" class="icon-16" />
      </div>
      <input
        class="s-checkbox__original"
        type="checkbox"
        :checked="isChecked"
        aria-hidden="false"
        :value="checkedValue"
        :disabled="state === 'disabled'"
        @change="handleModelChange"
      />
    </div>
    <div v-if="label || $slots.default" class="s-checkbox__label">
      <!--  @slot Слот для контента checkbox   -->
      <slot>{{ label }}</slot>
    </div>
  </label>
</template>

<script lang="ts">
import { Component, Vue, Model, Prop } from 'vue-property-decorator'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'
import type { SCheckboxValue } from '@/lib-components/SCheckbox/SCheckbox.type'

/** Компонент для выбора значения */
@Component({
  components: {
    SvgIcon,
  }
})
export default class SCheckbox extends Vue {
  @Model('change')
  value!: SCheckboxValue[] | SCheckboxValue | null

  /** Значение checkbox */
  @Prop()
  checkedValue?: SCheckboxValue

  /** Ярлык checkbox */
  @Prop(String)
  label?: string

  /** Возможность использовать один v-model на несколько checkbox */
  @Prop({ type: Boolean, default: false })
  multiple!: boolean

  /** Состояние чекбокса ("disabled", "error") */
  @Prop({ type: String, default: 'default' })
  state!: string

  get isChecked(): SCheckboxValue[] | SCheckboxValue | null {
    if (this.multiple && this.checkedValue !== undefined) {
      if (!Array.isArray(this.value)) return false

      return Array.isArray(this.value) && this.value.includes(this.checkedValue)
    }

    if (this.checkedValue === undefined) {
      return this.value
    }

    return this.checkedValue === this.value
  }

  handleModelChange(event: InputEvent): void {
    let input = this.value
    const element = event.target as HTMLInputElement

    if (this.multiple) {
      if (!Array.isArray(input)) {
        input = []
      }

      if (Array.isArray(input) && this.checkedValue !== undefined) {
        const index = input.indexOf(this.checkedValue)

        if (index !== -1) {
          input.splice(index, 1)
        } else {
          input.push(this.checkedValue)
        }
      }
    } else if (this.checkedValue !== undefined && element) {
      input = element.checked ? this.checkedValue : null
    } else if (element) {
      input = element.checked
    }

    this.$emit('change', input)
  }

  created(): void {
    if (this.multiple && !Array.isArray(this.value) && this.value) {
      this.$emit('change', [this.value])
    }
  }
}
</script>
