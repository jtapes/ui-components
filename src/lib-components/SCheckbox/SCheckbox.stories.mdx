import { action } from '@storybook/addon-actions'
import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest';
import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Canvas,
} from '@storybook/addon-docs'
import SCheckbox from '@/lib-components/SCheckbox/SCheckbox.vue'
import results from '/.jest-test-results.json';

<Meta
  title="DESIGN SYSTEM/Atom/Checkbox"
  component={SCheckbox}
  decorators={[withDesign, withTests({ results })]}
  argTypes={{
    label: { control: 'text' },
    state: {
      control: { type: 'select' },
      options: [
        'disabled',
        'error',
        'default'
      ]
    }
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/IIeFkxNoHGcaqNHlVeNrhW/MVP-3-Market-UI-Kit?node-id=846%3A1100',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# Checkbox

<Description of={SCheckbox} />

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SCheckbox },
  template:
    `<div class="d-f fd-c">
      <SCheckbox v-model="selectedValue" :label="label" :state="state" @change="handleChange"></SCheckbox>
      <SCheckbox v-model="selectedValue2">
        С <span class="link link--underline">типовыми условиями договора</span> ознакомлен
      </SCheckbox>
</div>
`,
	data() {
		return {
			selectedValue: null,
      selectedValue2: null
		}
	},
	methods: {
		handleChange() {
			action('change')()
		}
	}
})

<Canvas withToolbar withSource="none">
  <Story name="SCheckbox" args={{ label: 'Checkbox' }}>
    {Template.bind({})}
  </Story>
</Canvas>

## Пример

<Source
  language="html"
  code={`
    <s-checkbox v-model="selectedValue" :label="Checkbox"/>
  `}
/>

<ArgsTable of={SCheckbox} />
