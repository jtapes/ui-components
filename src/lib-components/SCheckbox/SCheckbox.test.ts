import { mount } from '@vue/test-utils'
import SCheckbox from '@/lib-components/SCheckbox/SCheckbox.vue'

describe('SCheckbox', () => {
  test('Is Vue snapshot', () => {
    const wrapper = mount(SCheckbox, {
      propsData: {
        value: 1,
      },
      stubs: ['svg-icon', 'nuxt-link'],
    })
    expect(wrapper.element).toMatchSnapshot()
  })
  test('Is emitted by $emit()', async () => {
    const wrapper = mount(SCheckbox, {
      propsData: {
        value: 1,
        multiple: true,
      },
      stubs: ['svg-icon', 'nuxt-link'],
    })

    const input = wrapper.find('input[type="checkbox"]')
    await input.setChecked((input.element as HTMLInputElement).checked)
    await input.trigger('change')
    expect(wrapper.emitted('change')).toBeTruthy()
  })
  test('Label of checkbox exist label prop && emit on not multiple value', async () => {
    const label = 'Checkbox'

    const wrapper = mount(SCheckbox, {
      propsData: {
        value: 1,
        checkedValue: 1,
        label,
      },
      stubs: ['svg-icon', 'nuxt-link'],
    })

    const labelContent = wrapper.find('.s-checkbox__label')
    expect(labelContent.text()).toBe(label)

    const input = wrapper.find('input[type="checkbox"]')
    await input.setChecked((input.element as HTMLInputElement).checked)
    await input.trigger('change')

    await wrapper.setProps({
      value: 1,
      checkedValue: undefined,
      label,
    })
    await input.setChecked((input.element as HTMLInputElement).checked)
    await input.trigger('change')

    expect(wrapper.emitted('change')).toBeTruthy()
  })
  test('multiple with array value', async () => {
    const wrapper = mount(SCheckbox, {
      propsData: {
        value: [1],
        multiple: true,
        checkedValue: 1,
      },
      stubs: ['svg-icon', 'nuxt-link'],
    })

    const input = wrapper.find('input[type="checkbox"]')
    await input.setChecked((input.element as HTMLInputElement).checked)
    await input.trigger('change')
    expect(wrapper.emitted('change')).toBeTruthy()
  })
})
