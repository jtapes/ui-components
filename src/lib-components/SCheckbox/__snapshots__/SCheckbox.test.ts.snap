// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SCheckbox Is Vue snapshot 1`] = `
<label
  class="s-checkbox text-sm s-checkbox--default s-checkbox--checked"
>
  <div
    class="s-checkbox__input"
  >
    <div
      class="s-checkbox__inner"
    >
      <svg-icon-stub
        class="icon-16"
        name="checked-16"
      />
    </div>
     
    <input
      aria-hidden="false"
      class="s-checkbox__original"
      type="checkbox"
      value=""
    />
  </div>
   
  <!---->
</label>
`;
